File: CSharp/Gravical.Poker.Core/HandGrouping.cs
using System;
using System.Collections.Generic;

namespace Gravical.Poker.Core
{
    /// <summary>
    ///  HandGrouping builds list of multiple cards, pair, three of a kind, and four of a kind.
    /// </summary>
    public class HandGrouping
    {
        public List<List<Card>> FourOfAKind { get; }
        public List<List<Card>> ThreeOfAKind { get; }
        public List<List<Card>> Pair { get; }
        public List<List<Card>> Single { get; }

        public HandGrouping(Card[] cards)
        {
            Guards.ArgumentNotNull(cards, nameof(cards));
            Guards.ArgumentSuccess(cards.Length >= 5 && cards.Length <= 7, nameof(cards), "Only 5-7 cards can be grouped");

            var groups = new Dictionary<Face, List<Card>>();
            var unique = new HashSet<Card>();
            foreach (var card in cards)
            {
                if (!card.IsValid)
                    throw new ArgumentException("Invalid cards cannot be grouped", nameof(cards));
                if (unique.Contains(card))
                    throw new ArgumentException("Duplicate cards cannot be grouped", nameof(cards));
                unique.Add(card);

                if (!groups.ContainsKey(card.Face))
                    groups.Add(card.Face, new List<Card>());

                groups[card.Face].Add(card);
            }

            FourOfAKind = new List<List<Card>>();
            ThreeOfAKind = new List<List<Card>>();
            Pair = new List<List<Card>>();
            Single = new List<List<Card>>();

            foreach (var face in groups.Keys)
            {
                var list = groups[face];
                if (list.Count == 4)
                    FourOfAKind.Add(list);
                else if (list.Count == 3)
                    ThreeOfAKind.Add(list);
                else if (list.Count == 2)
                    Pair.Add(list);
                else if (list.Count == 1)
                    Single.Add(list);
                else
                    throw new InvalidOperationException("Impossible grouping");
            }
        }
    }
}

File: JavaScript/src/HandGrouping.mjs
import { Card } from './Card.mjs';
import { Random } from './Random.mjs';

// HandGrouping builds list of multiple cards, pair, three of a kind, and four of a kind.
export class HandGrouping {

	constructor(cards)
	{
		if (!cards) throw "Cards is null";
		if (!(cards instanceof Array)) throw "Cards must be an array of cards";
		if (cards.length < 5 || cards.length > 7) throw "Only 5-7 cards can be grouped";

		this.fourOfAKind = [];
		this.threeOfAKind = [];
		this.pair = [];
		this.single = [];

		let groups = new Map();
		let unique = new Map();
		for (const card of cards) {
			if (!(card instanceof Card))
				throw "One or more initializers isn't a card"
			if (!card.isValid())
				throw "Invalid cards cannot be grouped";
			if (unique.has(card.toString()))
				throw "Duplicate cards cannot be grouped";

			unique.set(card.toString(), card);

			if (!groups.has(card.face))
				groups.set(card.face, []);

			groups.get(card.face).push(card);
		}

		for (const [face, list] of groups.entries()) {
			if (list.length == 4)
				this.fourOfAKind.push(list);
			else if (list.length == 3)
				this.threeOfAKind.push(list);
			else if (list.length == 2)
				this.pair.push(list);
			else if (list.length == 1)
				this.single.push(list);
			else
				throw "Impossible grouping";
		}
	}
}
